### ISSUE TITLE
A typo in ingress redirect regexp makes traefik pod fragile

### ISSUE THREAD

**Issue by kachkaev** [containous/traefik#3731]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I accidentally forgot to double-escape `.` in a redirect regex and created an ingress object like this:

```yaml
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    traefik.ingress.kubernetes.io/redirect-regex: ^https?://example\.com/(.*) # ← missing one more \ before .com
    traefik.ingress.kubernetes.io/redirect-replacement: https://www.example.com/$1
    traefik.ingress.kubernetes.io/redirect-permanent: "true"
  name: example
spec:
  rules:
  - host: example.com
    http:
      paths:
      - backend:
          serviceName: example
          servicePort: http
  - host: www.example.com
    http:
      paths:
      - backend:
          serviceName: example
          servicePort: http
```

This triggered the following warning:

```
level=error msg="Near line 295 (last key parsed 'frontends.example.com.redirect.regex'): invalid escape character '.'; only the following escape characters are allowed: \\b,\\t, \\n, \\f, \\r, \\\", \\\\, \\uXXXX, and \\UXXXXXXXX"
```

Searching for this message quickly brought me to https://github.com/containous/traefik/issues/3409#issuecomment-392827301 and I fixed my issue by adding one more `\` before the dot (`^https?://example\.com/(.*)` → `^https?://example\\.com/(.*)`). 👍 

However, I noticed that recreating the traefik's pod while an issue with a regexp is out there, makes it fail to apply _all existing ingress rules_. This hides the entire workload from the external users and people start seeing 404 for any URL as if there were no ingress rules at all. Interestingly, the new pod reports that it is healthy and there are no other signs of such a major problem. 

<details>
<summary>traefik pod logs</summary>

```
time="2018-08-05T15:33:01Z" level=info msg="Using TOML configuration file /config/traefik.toml"
time="2018-08-05T15:33:01Z" level=warning msg="web provider configuration is deprecated, you should use these options : api, rest provider, ping and metrics"
time="2018-08-05T15:33:01Z" level=info msg="Traefik version v1.7.0-rc3 built on 2018-08-01_01:37:51PM"
time="2018-08-05T15:33:01Z" level=info msg="\nStats collection is enabled.\nMany thanks for contributingto Traefik's improvement by allowing us to receive anonymous information from your configuration.\nHelp us improve Traefik by leaving this feature on :)\nMore details on: https://docs.traefik.io/basics/#collected-data\n"
time="2018-08-05T15:33:02Z" level=info msg="Preparing server https &{Address::443 TLS:0xc4202a3200 Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc4204fae80} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-08-05T15:33:03Z" level=info msg="Preparing server http &{Address::80 TLS:<nil> Redirect:0xc42060ae80 Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc4204faea0} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-08-05T15:33:03Z" level=info msg="Preparing server traefik &{Address::8080 TLS:<nil> Redirect:<nil> Auth:0xc420392e70 WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc4204faec0} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-08-05T15:33:03Z" level=info msg="Starting provider configuration.ProviderAggregator {}"
time="2018-08-05T15:33:03Z" level=info msg="Starting server on :443"
time="2018-08-05T15:33:03Z" level=info msg="Starting server on :80"
time="2018-08-05T15:33:03Z" level=info msg="Starting server on :8080"
time="2018-08-05T15:33:03Z" level=info msg="Starting provider *kubernetes.Provider {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"\",\"Token\":\"\",\"CertAuthFilePath\":\"\",\"DisablePassHostHeaders\":false,\"EnablePassTLSCert\":false,\"Namespaces\":null,\"LabelSelector\":\"\",\"IngressClass\":\"\",\"IngressEndpoint\":null}"
time="2018-08-05T15:33:03Z" level=info msg="Starting provider *acme.Provider {\"Email\":\"letsencrypt@example.com\",\"ACMELogging\":false,\"CAServer\":\"https://acme-v02.api.letsencrypt.org/directory\",\"Storage\":\"/acme/acme.json\",\"EntryPoint\":\"https\",\"KeyType\":\"\",\"OnHostRule\":true,\"OnDemand\":false,\"DNSChallenge\":null,\"HTTPChallenge\":{\"EntryPoint\":\"http\"},\"TLSChallenge\":null,\"Domains\":null,\"Store\":{}}"
time="2018-08-05T15:33:03Z" level=info msg="Testing certificate renew..."
time="2018-08-05T15:33:03Z" level=info msg="ingress label selector is: \"\""
time="2018-08-05T15:33:03Z" level=info msg="Creating in-cluster Provider client"
time="2018-08-05T15:33:03Z" level=error msg="Near line 291 (last key parsed 'frontends.example.com.redirect.regex'): invalid escape character '.'; only the following escape characters are allowed: \\b,\\t, \\n, \\f, \\r, \\\", \\\\, \\uXXXX, and \\UXXXXXXXX"
time="2018-08-05T15:33:06Z" level=info msg="Server configuration reloaded on :8080"
time="2018-08-05T15:33:06Z" level=info msg="Server configuration reloaded on :443"
time="2018-08-05T15:33:06Z" level=info msg="Server configuration reloaded on :80"
time="2018-08-05T15:33:16Z" level=error msg="Near line 295 (last key parsed 'frontends.example.com.redirect.regex'): invalid escape character '.'; only the following escape characters are allowed: \\b,\\t, \\n, \\f, \\r, \\\", \\\\, \\uXXXX, and \\UXXXXXXXX"
```

</details>

  
Fixing the faulty regexp after the pod has restarted does not ‘heal’ ingress rules, so the only remedy is to fix the regexp, kill the pod and wait for the new one to launch.

### What did you expect to see?

I expected traefik to be a bit more tolerant to a missed `\` in one of the annotations and not make all ingress rules unavailable on pod recreation. This behaviour may lead to a pretty long downtime if traefik restarts in the middle of the night when no one is around.

The only reason why I could quickly detect and fix the problem in regex for myself was because I was playing with a new cluster. Originally, my annotation looked like this and it worked:

```yaml
traefik.ingress.kubernetes.io/redirect-regex: ^https?://example.com/(.*)
```

I remembered that `.` stands for any character and changed the annotation to:

```yaml
traefik.ingress.kubernetes.io/redirect-regex: ^https?://example\.com/(.*) # \. instead of \\.
```

**All worked!** (I guess it's because traefik kept the last known valid rule in its memory)

I can easily imagine someone else doing this kind of change and then leaving traefik working for days or weeks. Then, once traefik dies at one point and a new pod gets created, all ingress rules go down! 😬 

It'd be great if a typo in a regex only affected the rule it belongs to.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
1.7.0-rc3-alpine
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

official k8s helm chart

**Comment by dtomcej** [comment]
The regex in the kubernetes provider does not actually parse the regex to confirm it is valid:
https://github.com/containous/traefik/blob/master/provider/kubernetes/kubernetes.go#L885
It just passes the value into the template.

Then when the template is submitted to the server, it causes the unparseable template.

The solution would be to parse the regex to ensure that it is valid before continuing.

**Comment by kachkaev** [comment]
If my recent tests were adequate, a single typo done once in a regexp makes it necessary to restart the pod. Otherwise, the rules are ‘frozen’ at the last error-free state and fixing the faulty ingress does not help.

**Comment by dtomcej** [comment]
@kachkaev let me rephrase:

The solution would be to _have traefik parse the regex_ from the annotations to ensure that it is valid before _continuing to apply the regex to the template_. This will allow the ingress to be skipped, as is expected from a mal-formed ingress/annotation`

😄

**Comment by traefiker** [comment]
Closed by #3753.
